### PR TITLE
Cuddly Chameleons

### PR DESCRIPTION
**Pull request by ks129**
Cuddly Chameleons team project PR.

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
No merge issues.

No flake8 linting errors.

File: cuddly-chameleons/cuddly_chameleons/retro_news/migrations/0003_auto_20200808_1230.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Generated by Django 3.1 on 2020-08-08 12:30
```
Comment (lemonsaurus) [inline cuddly-chameleons/cuddly_chameleons/retro_news/migrations/0003_auto_20200808_1230.py:1]: It's fine to use Django to generate commits, but you should rename them when they get ugly names like this. `0003_auto_20200808_1230.py` gives me absolutely no clues what this migration does. Autogenerate them, then rename them. They'll still work.

File: cuddly-chameleons/cuddly_chameleons/retro_news/models.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+from django.contrib.auth.models import AbstractUser
+from django.db import models
+
+
+class CustomUser(AbstractUser):
+    """Extend default user for case when we need to add something to this."""
+
+    pass
```
Comment (lemonsaurus) [inline cuddly-chameleons/cuddly_chameleons/retro_news/models.py:8]: Did you know that this `pass` doesn't need to be here? I don't really mind that it is there, but it is possible to have a class which contains nothing as long as it has a docstring. The docstring counts as sufficient content.

File: cuddly-chameleons/cuddly_chameleons/retro_news/models.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+class BlogArticle(models.Model):
+    """Object for blog post."""
+
+    title = models.CharField(max_length=100, unique=True)
+    content = models.TextField()
+    author = models.ForeignKey(CustomUser, on_delete=models.CASCADE)
+    created = models.DateTimeField(auto_now=True)
+
+    def __str__(self):
+        return "<BlogArticle: {}>".format(self.title)
```
Comment (lemonsaurus) [inline cuddly-chameleons/cuddly_chameleons/retro_news/models.py:20]: Why not use an f-string here?

**Review by lemonsaurus — COMMENTED** [review]
Thanks for the pull request, Cuddly Chameleons, and thank you so much for taking part in this event! 

## Readme
Your readme is pretty good, including a project description, setup instructions, how to use the app, and listing both team members and critical dependencies.

I would've liked to see more information about what makes your submission thematic, and maybe screenshots from the app in action. I'm also curious why there isn't any licensing information - did you create all the assets for this project yourself, or did you forget to credit the assets? Remember, just because an asset is free doesn't mean you shouldn't credit it. Giving credit (even when not required) is considered good etiquette in the open-source world.


## Project stack
You're using `docker`, `docker-compose`,  and `pipenv` to manage environments and dependencies - some of my favorite tools. You've even included an Azure DevOps CI script for this project! A very good idea to manage environments and dependencies when you're going to share your project with someone, like if it's open-source or it's going before a panel of judges. It makes it easier for us, and for you, and ensures that we don't run into weird problems that you didn't see on your end.

Instead of running the built-in django server, you've added `gunicorn` as a dependency, which is a very good idea for projects like this. This helps make the app more scalable, and protect it from DDoS vulnerabilities.

For your frontend, you've chosen to build an SPA with React, which works very well together with Django and makes for an extremely powerful app that you can basically build anything you can imagine with.

Overall, I think your choice of project stack is very impressive, and can stand toe-to-toe with most industry standard professional projects.


## Use of Python and Django
It's very nice to see you using Whitenoise, my favorite plugin for managing static files. I wish your migration files had better names - it's fine to use Django to autogenerate the migrations, but you should rename those files so they're not called stuff like `0003_auto_20200808_1230.py`. The migration filename should describe what the migration does.

Your Python code is idiomatic and modern. It makes use of features like f-strings, special methods for classes, and type annotations. There are some instances of .format() interpolation, which I don't think are adequately justified. It's fine to use .format if you need to interpolate it later (for example), but if it's going to be interpolated immediately, there's really no reason not to use f-strings these days.

Overall, this project makes very good use of both Python and Django. Very nice work.


## Commit quality
Your commit quality is good - almost all the commits describe what they do, and they stay inside the 50 character limit. However, almost none of the commits have commit message body. For the larger commits, it would be a good idea to write a good body for the commit where you explain what you've done, why you've done it, and how you did it. Take a look at [this article](https://fatbusinessman.com/2019/my-favourite-git-commit) for more information on precisely what a great commit should contain.

I also wish this pull request had a good description. It's a good habit to write beautiful pull requests.


# Things you did well
- Good commit quality.
- Good readme.
- Very good use of Django
- Mostly modern and idiomatic Python code.
- World-class project stack. Lots of gold standard components.

# Things that could be better
- Be more consistent about how you interpolate strings.
- Where's the licensing information for your assets?
- Write a good pull request description.
- Don't be afraid to use the body of your commit message to provide more context on not only what you did, but why you did it.
- It looks like @ks129 committed over 90% of the code for this project. I'm worried that you guys did not work well together as a team - can you comment on this? Just leave a reply to this review.

**Comment by ks129** [comment]
@lemonsaurus About this teamwork, problem was mess in team (team got merged with another team when there was few days left, because 3 members dropped out from team, and then we got 1 more member) and also skill of other team members. They was new on Django/React, but they helped test it (like for example they found this bug that allow all posts fetching only in Safari) and few more smaller bugs. About Safari-only bug, I started working on it too late: only few hours was left and finally I was unable to find solution to it. I had plan to write PR description, but I searched solution to this bug until last minute.

That `pass` is not required on class when there is docstring was new for me. About missing f-string, looks like I missed it during PR reviewing. 😅 and I totally forget about migrations naming.

**Comment by lemonsaurus** [comment]
Okay - if the team members were helpful to you in testing and bugfixing, I think that's good enough. 

It's a shame about the Safari-only bug. I doubt most of us use Safari, but I guess we'll have to install it to test your app.

**Comment by ks129** [comment]
@lemonsaurus This is not 100% required to use Safari, just when you are not in Safari, you have to log in to see posts and comments, because other browsers add trailing slash what require authentication in Django side.

**Pull request merged**